src/raftstore/store/msg.rs

```diff
@@ -18,11 +18,83 @@ use std::fmt;
 use kvproto::raft_serverpb::RaftMessage;
 use kvproto::raft_cmdpb::{RaftCmdRequest, RaftCmdResponse};
 use kvproto::metapb::RegionEpoch;
+
 use raft::SnapshotStatus;
 use util::escape;
 
-pub type Callback = Box<FnBox(RaftCmdResponse) + Send>;
-pub type BatchCallback = Box<FnBox(Vec<Option<RaftCmdResponse>>) + Send>;
+use super::engine::Snapshot;
+
+#[derive(Debug)]
+pub struct ReadResponse {
+    pub response: RaftCmdResponse,
+    pub snapshot: Option<Snapshot>,
+}
+
+#[derive(Debug)]
+pub struct WriteResponse {
+    pub response: RaftCmdResponse,
+}
+
+pub type ReadCallback = Box<FnBox(ReadResponse) + Send>;
+pub type WriteCallback = Box<FnBox(WriteResponse) + Send>;
+pub type BatchReadCallback = Box<FnBox(Vec<Option<ReadResponse>>) + Send>;
+
+/// Variants of callbacks for `Msg`.
+///  - `Read`: a callbak for read only requests including `StatusRequest`,
+///         `GetRequest` and `SnapRequest`
+///  - `Write`: a callback for write only requests including `AdminRequest`
+///          `PutRequest`, `DeleteRequest` and `DeleteRangeRequest`.
+///  - `BatchRead`: cllbacks for a batch read request.
+pub enum Callback {
+    /// No callback.
+    None,
+    /// Read callback.
+    Read(ReadCallback),
+    /// Write callback.
+    Write(WriteCallback),
+    // Batch read callbacks.
+    BatchRead(BatchReadCallback),
+}
+
+impl Callback {
+    pub fn invoke_with_response(self, resp: RaftCmdResponse) {
```

Comment:
It's more clear to define `invoke_write` instead.

Reply:
It's meat to works with `None`, `Read` and `Write`.